File: classified/packages/plugin-vision/src/action-audio.ts
// Audio capture actions for the vision plugin
import {
  type Action,
  type ActionExample,
  ActionResult,
  type HandlerCallback,
  type IAgentRuntime,
  type Memory,
  type State,
  logger,
} from '@elizaos/core';
import { VisionService } from './service';

export const transcribeAudioAction: Action = {
  name: 'TRANSCRIBE_AUDIO',
  similes: ['LISTEN', 'HEAR_AUDIO', 'CAPTURE_AUDIO', 'RECORD_AUDIO'],
  description:
    'Captures audio from the microphone and transcribes it to text. Useful for voice commands, dictation, or understanding spoken content.',
  validate: async (runtime: IAgentRuntime, _message: Memory, _state?: State): Promise<boolean> => {
    // Check if microphone is enabled in runtime settings
    const microphoneEnabled =
      runtime.getSetting('ENABLE_MICROPHONE') === 'true' ||
      runtime.getSetting('VISION_MICROPHONE_ENABLED') === 'true';

    if (!microphoneEnabled) {
      logger.debug('[transcribeAudioAction] Microphone disabled in settings.');
      return false;
    }

    const visionService = runtime.getService<VisionService>('VISION');
    return !!visionService && visionService.isActive();
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: any,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult> => {
    try {
      const visionService = runtime.getService<VisionService>('VISION');
      if (!visionService) {
        logger.error('[transcribeAudioAction] VisionService not found');
        return {
          success: false,
          error: 'Vision service not available',
        };
      }

      // Check if audio capture is available
      const audioCapture = (visionService as any).audioCapture;
      const streamingAudioCapture = (visionService as any).streamingAudioCapture;

      if (!audioCapture && !streamingAudioCapture) {
        logger.error('[transcribeAudioAction] No audio capture service available');
        return {
          success: false,
          error: 'Audio capture not initialized. Please enable microphone in settings.',
        };
      }

      logger.info('[transcribeAudioAction] Starting audio transcription...');

      // Use batch audio capture for a single transcription
      if (audioCapture) {
        const transcription = await audioCapture.recordAndTranscribe();

        if (transcription) {
          logger.info(`[transcribeAudioAction] Transcribed: "${transcription}"`);

          if (callback) {
            await callback({
              text: `I heard you say: "${transcription}"`,
              source: 'transcribeAudioAction',
            });
          }

          return {
            success: true,
            data: {
              transcription,
              timestamp: new Date().toISOString(),
            },
          };
        } else {
          return {
            success: false,
            error: 'No speech detected or transcription failed',
          };
        }
      }

      return {
        success: false,
        error: 'Audio capture service not properly initialized',
      };
    } catch (error) {
      logger.error('[transcribeAudioAction] Error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred',
      };
    }
  },
  examples: [
    [
      {
        name: 'elizaos',
        content: {
          text: 'Listen to what I have to say',
          action: 'TRANSCRIBE_AUDIO',
        },
      },
    ],
    [
      {
        name: 'elizaos',
        content: {
          text: 'Can you hear me?',
          action: 'LISTEN',
        },
      },
    ],
    [
      {
        name: 'elizaos',
        content: {
          text: 'Record audio and tell me what you hear',
          action: 'CAPTURE_AUDIO',
        },
      },
    ],
  ] as ActionExample[][],
};

export const toggleStreamingAudioAction: Action = {
  name: 'TOGGLE_STREAMING_AUDIO',
  similes: ['START_LISTENING', 'STOP_LISTENING', 'CONTINUOUS_LISTENING'],
  description:
    'Toggles continuous audio streaming with voice activity detection. When enabled, the agent will continuously listen and transcribe speech.',
  validate: async (runtime: IAgentRuntime, _message: Memory, _state?: State): Promise<boolean> => {
    const microphoneEnabled =
      runtime.getSetting('ENABLE_MICROPHONE') === 'true' ||
      runtime.getSetting('VISION_MICROPHONE_ENABLED') === 'true';

    const streamingEnabled = runtime.getSetting('USE_STREAMING_AUDIO') === 'true';

    return microphoneEnabled && streamingEnabled;
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: any,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult> => {
    try {
      const visionService = runtime.getService<VisionService>('VISION');
      if (!visionService) {
        return {
          success: false,
          error: 'Vision service not available',
        };
      }

      const streamingAudioCapture = (visionService as any).streamingAudioCapture;
      if (!streamingAudioCapture) {
        return {
          success: false,
          error: 'Streaming audio not initialized. Enable USE_STREAMING_AUDIO in settings.',
        };
      }

      const isCapturing = streamingAudioCapture.isCapturing;

      if (isCapturing) {
        await streamingAudioCapture.stop();
        logger.info('[toggleStreamingAudioAction] Stopped streaming audio capture');

        if (callback) {
          await callback({
            text: 'I have stopped listening continuously.',
            source: 'toggleStreamingAudioAction',
          });
        }

        return {
          success: true,
          data: {
            streaming: false,
            message: 'Streaming audio capture stopped',
          },
        };
      } else {
        await streamingAudioCapture.start();
        logger.info('[toggleStreamingAudioAction] Started streaming audio capture');

        if (callback) {
          await callback({
            text: 'I am now listening continuously. I will transcribe everything I hear.',
            source: 'toggleStreamingAudioAction',
          });
        }

        return {
          success: true,
          data: {
            streaming: true,
            message: 'Streaming audio capture started',
          },
        };
      }
    } catch (error) {
      logger.error('[toggleStreamingAudioAction] Error:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred',
      };
    }
  },
  examples: [
    [
      {
        name: 'elizaos',
        content: {
          text: 'Start listening continuously',
          action: 'START_LISTENING',
        },
      },
    ],
    [
      {
        name: 'elizaos',
        content: {
          text: 'Stop continuous listening',
          action: 'STOP_LISTENING',
        },
      },
    ],
  ] as ActionExample[][],
};

File: cpp_merge/classified/packages/plugin-vision/src/action-audio.cpp
#include "action-audio.h"

std::shared_ptr<Action> transcribeAudioAction = object{
    object::pair{std:("name"), std:("TRANSCRIBE_AUDIO")}, 
    object::pair{std:("similes"), array<string>{ std:("LISTEN"), std:("HEAR_AUDIO"), std:("CAPTURE_AUDIO"), std:("RECORD_AUDIO") }}, 
    object::pair{std:("description"), std:("Captures audio from the microphone and transcribes it to text. Useful for voice commands, dictation, or understanding spoken content.")}, 
    object::pair{std:("validate"), [=](auto runtime, auto _message, auto _state = undefined) mutable
    {
        auto microphoneEnabled = OR((runtime->getSetting(std:("ENABLE_MICROPHONE")) == std:("true")), (runtime->getSetting(std:("VISION_MICROPHONE_ENABLED")) == std:("true")));
        if (!microphoneEnabled) {
            logger->debug(std:("[transcribeAudioAction] Microphone disabled in settings."));
            return false;
        }
        auto visionService = runtime->getService<std::shared_ptr<VisionService>>(std:("VISION"));
        return AND((!!visionService), (visionService->isActive()));
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto _state = undefined, auto _options = undefined, auto callback = undefined, auto _responses = undefined) mutable
    {
        try
        {
            auto visionService = runtime->getService<std::shared_ptr<VisionService>>(std:("VISION"));
            if (!visionService) {
                logger->error(std:("[transcribeAudioAction] VisionService not found"));
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("error"), std:("Vision service not available")}
                };
            }
            auto audioCapture = (as<any>(visionService))["audioCapture"];
            auto streamingAudioCapture = (as<any>(visionService))["streamingAudioCapture"];
            if (AND((!audioCapture), (!streamingAudioCapture))) {
                logger->error(std:("[transcribeAudioAction] No audio capture service available"));
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("error"), std:("Audio capture not initialized. Please enable microphone in settings.")}
                };
            }
            logger->info(std:("[transcribeAudioAction] Starting audio transcription..."));
            if (audioCapture) {
                auto transcription = std::async([=]() { audioCapture["recordAndTranscribe"](); });
                if (transcription) {
                    logger->info(std:("[transcribeAudioAction] Transcribed: "") + transcription + std:("""));
                    if (callback) {
                        std::async([=]() { callback(object{
                            object::pair{std:("text"), std:("I heard you say: "") + transcription + std:(""")}, 
                            object::pair{std:("source"), std:("transcribeAudioAction")}
                        }); });
                    }
                    return object{
                        object::pair{std:("success"), true}, 
                        object::pair{std:("data"), object{
                            object::pair{std:("transcription"), std:("transcription")}, 
                            object::pair{std:("timestamp"), ((std::make_shared<Date>()))->toISOString()}
                        }}
                    };
                } else {
                    return object{
                        object::pair{std:("success"), false}, 
                        object::pair{std:("error"), std:("No speech detected or transcription failed")}
                    };
                }
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), std:("Audio capture service not properly initialized")}
            };
        }
        catch (const any& error)
        {
            logger->error(std:("[transcribeAudioAction] Error:"), error);
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), (is<Error>(error)) ? any(error->message) (std:("Unknown error occurred"))}
            };
        }
    }
    }, 
    object::pair{std:("examples"), as<array<array<std::shared_ptr<ActionExample>>>>(array<array<std::shared_ptr<ActionExample>>>{ array<object>{ object{
        object::pair{std:("name"), std:("elizaos")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Listen to what I have to say")}, 
            object::pair{std:("action"), std:("TRANSCRIBE_AUDIO")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("elizaos")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Can you hear me?")}, 
            object::pair{std:("action"), std:("LISTEN")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("elizaos")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Record audio and tell me what you hear")}, 
            object::pair{std:("action"), std:("CAPTURE_AUDIO")}
        }}
    } } })}
};
std::shared_ptr<Action> toggleStreamingAudioAction = object{
    object::pair{std:("name"), std:("TOGGLE_STREAMING_AUDIO")}, 
    object::pair{std:("similes"), array<string>{ std:("START_LISTENING"), std:("STOP_LISTENING"), std:("CONTINUOUS_LISTENING") }}, 
    object::pair{std:("description"), std:("Toggles continuous audio streaming with voice activity detection. When enabled, the agent will continuously listen and transcribe speech.")}, 
    object::pair{std:("validate"), [=](auto runtime, auto _message, auto _state = undefined) mutable
    {
        auto microphoneEnabled = OR((runtime->getSetting(std:("ENABLE_MICROPHONE")) == std:("true")), (runtime->getSetting(std:("VISION_MICROPHONE_ENABLED")) == std:("true")));
        auto streamingEnabled = runtime->getSetting(std:("USE_STREAMING_AUDIO")) == std:("true");
        return AND((microphoneEnabled), (streamingEnabled));
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto _state = undefined, auto _options = undefined, auto callback = undefined, auto _responses = undefined) mutable
    {
        try
        {
            auto visionService = runtime->getService<std::shared_ptr<VisionService>>(std:("VISION"));
            if (!visionService) {
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("error"), std:("Vision service not available")}
                };
            }
            auto streamingAudioCapture = (as<any>(visionService))["streamingAudioCapture"];
            if (!streamingAudioCapture) {
                return object{
                    object::pair{std:("success"), false}, 
                    object::pair{std:("error"), std:("Streaming audio not initialized. Enable USE_STREAMING_AUDIO in settings.")}
                };
            }
            auto isCapturing = streamingAudioCapture["isCapturing"];
            if (isCapturing) {
                std::async([=]() { streamingAudioCapture["stop"](); });
                logger->info(std:("[toggleStreamingAudioAction] Stopped streaming audio capture"));
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("I have stopped listening continuously.")}, 
                        object::pair{std:("source"), std:("toggleStreamingAudioAction")}
                    }); });
                }
                return object{
                    object::pair{std:("success"), true}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("streaming"), false}, 
                        object::pair{std:("message"), std:("Streaming audio capture stopped")}
                    }}
                };
            } else {
                std::async([=]() { streamingAudioCapture["start"](); });
                logger->info(std:("[toggleStreamingAudioAction] Started streaming audio capture"));
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("I am now listening continuously. I will transcribe everything I hear.")}, 
                        object::pair{std:("source"), std:("toggleStreamingAudioAction")}
                    }); });
                }
                return object{
                    object::pair{std:("success"), true}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("streaming"), true}, 
                        object::pair{std:("message"), std:("Streaming audio capture started")}
                    }}
                };
            }
        }
        catch (const any& error)
        {
            logger->error(std:("[toggleStreamingAudioAction] Error:"), error);
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), (is<Error>(error)) ? any(error->message) (std:("Unknown error occurred"))}
            };
        }
    }
    }, 
    object::pair{std:("examples"), as<array<array<std::shared_ptr<ActionExample>>>>(array<array<std::shared_ptr<ActionExample>>>{ array<object>{ object{
        object::pair{std:("name"), std:("elizaos")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Start listening continuously")}, 
            object::pair{std:("action"), std:("START_LISTENING")}
        }}
    } }, array<object>{ object{
        object::pair{std:("name"), std:("elizaos")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Stop continuous listening")}, 
            object::pair{std:("action"), std:("STOP_LISTENING")}
        }}
    } } })}
};

void Main(void)
{
}

MAIN
